### PR TITLE
Facilitate development of bot development framework (aderyn_pilot) and testing toolkit included

### PR DESCRIPTION
**Pull request by TilakMaddy**
😊 

```bash

Usage: aderyn_pilot <COMMAND>

Commands:
  new       Argument - Bot name (including path). For example if the name is `smart_bot`, you run "aderyn_pilot new path/to/smart_bot"
  generate  Argument - Detector name (including path). For example if the name is "unindexed_events", you run "aderyn_pilot generate path/to/smart_bot/bot_starter_pack/src/unindexed_events"
  help      Print this message or the help of the given subcommand(s)

Options:
  -h, --help     Print help
  -V, --version  Print version


```

It is not perfect has some todos and maybe even some bugs but tries to cover the common use cases like create bot and generate detectors verifiable by CI/CD. Please try it out and let me also know how you feel 😇

My idea was that there will be 3 kinds of users 

type 1 - basic user. Only uses `aderyn` command line
type 2 - advanced user. He creates his bot's own detectors and combines them with carefully chosen core detectors. (This would also require being able to test the custom detectors - which is why I have included the toolkit :P)
type 3 - does both of the above + submit the custom detectors to online contest portals. so this is where we need to track metadata

Since type 3 can feel like invading privacy - I plan to roll it out as an explicitly opt-in `feature=["track-metadata"]` for `aderyn_pilot`. So there will be some kind of consent. 

This PR doesn't cover type 3 users. but it has some provisions to facilitate the development in that direction

### CODE REVIEW

**Comment by TilakMaddy** [comment]
While this thing works - basic features are complete (new + generate commands) it is not in a healthy state for development - so I am working on creating a bot-creation-framework-assembler . It may take me some time but can save time in future 

That's the reason why it is still in draft. 

Thank you for reading : )

**Comment by TilakMaddy** [comment]
This is my 3rd take haha :P  and I am still not very satisfied ...but trying to get it out. It's low key embarrassing how bad this still is

**Comment by TilakMaddy** [comment]
Hey @alexroan I have added the following to workflow yml
```yaml

 bots:
    name: Check Bot creation and testing
    runs-on: ubuntu-latest
    steps:
      - name: Checkout sources
        uses: actions/checkout@v2

      - name: Install stable toolchain
        uses: actions-rs/toolchain@v1
        with:
          profile: minimal
          toolchain: stable
          override: true

      - name: Ask assembler make a dev archive.zip
        uses: actions-rs/cargo@v1
        with:
          command: run
          args: --bin bot_fw_assembler -- dev ../../aderyn_driver

      - name: Create a bot with pilot
        uses: actions-rs/cargo@v1
        with:
          command: run
          args: --bin aderyn_pilot -- new bot_ci_cd

      - name: Build the bot 
        uses: actions-rs/cargo@v1
        working-directory: bot_ci_cd/bot_starter_pack
        with:
          command: build

```
It says startup failure. Not able to figure out why. Can you help?

**Comment by TilakMaddy** [comment]
(I have tested locally all the steps that I have included in the yml file)

File: bot_starter_pack/src/config_tests.rs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        // Define your targets here
+    ]
+    .into()
+}
+
+/*  Example
+
+pub fn tests_configuration() -> TestsConfig {
+    vec![
+        TestsTarget::new("../TEST-CONTRACTS/out/xyz.sol/xyz.json")
```
Comment (TilakMaddy) [inline bot_starter_pack/src/config_tests.rs:17]: This is the "testing toolkit" - You specify a list of test files and list out all the detectors that you want to test against them. (here it is Example and AnotherExample detectors)

When `cargo test` is ran, the detectors will receive their respective list of test files (so relationship inversion is taken care of)

File: aderyn_pilot/templates/detector_rs.hbs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    use crate::config_tests::tests_configuration;
+
+    use super::{{detector_name_title_case}};
+
+    use aderyn_driver::context::loader::ContextLoader;
+    use aderyn_driver::detector::detector_test_helpers::load_contract;
+    use aderyn_driver::detector::Detector;
+
+    fn test_{{detector_name_camel_case}}_for(
```
Comment (TilakMaddy) [inline aderyn_pilot/templates/detector_rs.hbs:51]: this test function will run against all the test files that have this detector in their `with`  clause

**Review by alexroan — APPROVED** [review]
Excellent work @TilakMaddy.

I'm going to work on some follow-up actions after this is merged into `dev,` including:
- Renaming `aderyn_pilot` to something catchier
- Renaming `bot_starter_pack`
- DevEx things inside the detector workspace (like a foundry_workspace)

**Pull request merged**